src/foam/nanos/auth/SystemAuthService.js

```diff
@@ -11,17 +11,35 @@ foam.CLASS({
 
   javaImports: [
     'foam.nanos.auth.Group',
+    'foam.nanos.session.Session',
     'javax.security.auth.AuthPermission'
   ],
 
+  constants: [
+    { name: 'CACHE_KEY', value: 'system.auth.cache' }
+  ],
+
   methods: [
     {
       name: 'check',
       javaCode: `
         User user = ((Subject) x.get("subject")).getUser();
         Group group = (Group) x.get("group");
-        boolean isSystem = user != null && user.getId() == foam.nanos.auth.User.SYSTEM_USER_ID;
-        return isSystem || group != null && group.implies(x, new AuthPermission("*")) || getDelegate().check(x, permission);
+        if ( user != null && user.getId() == foam.nanos.auth.User.SYSTEM_USER_ID ) {
+          return true;
+        }
+        // Response from group implies is cached to maintain intended performance.
+        if ( group != null ) {
+          Session session = x.get(Session.class);
+          Boolean isSuper = (Boolean) session.getContext().get(CACHE_KEY);
+          if ( isSuper == null ) {
+            isSuper = group.implies(x, new AuthPermission("*"));
+            session.setContext(session.getContext().put(CACHE_KEY, isSuper));
+          }
+          return isSuper || getDelegate().check(x, permission);
```

Comment:
This seems fragile if the user changes their group. Could we just keep an instance variable with a ConcurrentHashMap to cache super groups?

Reply:
> This seems fragile if the user changes their group.

Can you elaborate your concern?

> Could we just keep an instance variable with a ConcurrentHashMap to cache super groups?

I don't know - When a request to the server is made does the system create a new instance of each java class exclusive to the user and is maintained during their session?